gemini/reasoning-engine/tutorial_graph_rag.ipynb

```diff
@@ -0,0 +1,1707 @@
+    "    for sub_list in temp:\n",
+    "        table_name = sub_list[0]\n",
+    "        col_name = (sub_list[0]).lower()\n",
+    "        value = sub_list[1]\n",
+    "        row_ct = transaction.execute_update(\n",
+    "            f\"INSERT INTO {table_name} ({col_name}) VALUES (@value)\",\n",
+    "            params={\"value\": value},\n",
+    "            param_types={\"value\": spanner.param_types.STRING}  # Adjust type if needed\n",
+    "        )\n",
+    "        print(\"{} record(s) inserted.\".format(row_ct))\n",
+    "\n",
+    "    for sub_list2 in temp2:\n",
+    "        table_name = sub_list2[0]\n",
+    "        col_name1 = ((sub_list2[1]))[0:1] + (\"1_Name\" if table_name == \"INFLUENCED_BY\" or table_name == \"LOCATED_IN\" else \"_Name\")\n",
+    "        value1 = sub_list2[2]\n",
+    "        col_name2 = ((sub_list2[3]))[0:1] + (\"2_Name\" if table_name == \"INFLUENCED_BY\" or table_name == \"LOCATED_IN\" else \"_Name\")\n",
+    "        value2 = sub_list2[4]\n",
+    "        print(table_name, col_name1, value1, col_name2, value2)\n",
+    "\n",
+    "        row_ct1 = transaction.execute_update(\n",
+    "            f\"INSERT INTO {table_name} ({col_name1}, {col_name2}) VALUES (@value1, @value2)\",\n",
+    "            params={\"value1\": value1, \"value2\": value2},\n",
+    "            param_types={\"value1\": spanner.param_types.STRING, \"value2\": spanner.param_types.STRING}  # Adjust types if needed\n",
+    "        )\n",
+    "        print(\"{} record(s) inserted.\".format(row_ct1))  # Print the result here\n",
+    "    i = 0\n",
+    "    for sub_list in TextEmbeddings:\n",
+    "        table_name = \"KgNode\"\n",
+    "        col_name1 = \"docID\"\n",
+    "        col_name2 = \"Name\"\n",
+    "        col_name3 = \"Doc\"\n",
+    "        col_name4 = \"DocEmbedding\"\n",
+    "        d = str(Texts[i])\n",
+    "        value1 = i + 1\n",
+    "        i = i + 1\n",
+    "        value2 = \"Lawrence Edward Page\"\n",
+    "        value3 = d\n",
+    "        value4 = sub_list\n",
+    "        print(col_name1, col_name2, col_name3, col_name4, value1, value2, value3, value4)\n",
+    "        row_ct1 = transaction.execute_update(\n",
+    "            f\"INSERT INTO {table_name} ({col_name1}, {col_name2}, {col_name3}, {col_name4}) VALUES (@value1, @value2, @value3, @value4)\",\n",
+    "            params={\"value1\": value1, \"value2\": value2, \"value3\": value3, \"value4\" : value4},\n",
+    "            param_types={\"value1\": spanner.param_types.INT64, \"value2\": spanner.param_types.STRING, \"value3\": spanner.param_types.STRING, \"value4\": spanner.param_types.Array(spanner.param_types.FLOAT64)})  # Adjust types if needed\n",
+    "\n",
+    "        print(\"{} record(s) inserted.\".format(row_ct1))\n",
+    "\n",
+    "# print(insert_values)  # This just prints the function object, remove this line\n",
+    "database.run_in_transaction(insert_values)"
```

Comment:
This database insert seems very complicated. Is there a simpler way to notate all of this?